libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1005 @@
+      auto primary_key_found = internal_use_do_not_use::kv_get(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), value_size);
+
+      if (!primary_key_found) {
+         return;
+      }
+
+      for (const auto& idx : secondary_indices) {
+         auto sec_tbl_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(value));
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size());
+      }
+
+      internal_use_do_not_use::kv_erase(db_name, contract_name.value, tbl_key.data(), tbl_key.size());
+   }
+
+protected:
+   kv_table() = default;
+
+   template <typename... Indices>
+   void setup_indices(uint64_t index_name, bool is_named, null_index* index, Indices... indices) {
+      ++index_name;
+      if constexpr (sizeof...(indices) > 0) {
+         setup_indices(index_name, is_named, indices...);
+      }
+   }
+
+   template <typename I, typename... Indices>
+   void setup_indices(uint64_t index_name, bool is_named, I index, Indices... indices) {
+      if (is_named) {
+         eosio::check(index->index_name.value > 0, "All indices must be named if one is named.");
+      } else {
+         eosio::check(index->index_name.value <= 0, "All indices must be named if one is named.");
+         index->index_name = eosio::name{index_name};
+      }
+
+      index->contract_name = contract_name;
+      index->table_name = table_name;
+      index->tbl = this;
+
+      index->setup();
+      secondary_indices.push_back(index);
+      ++index_name;
+      setup_indices(index_name, is_named, indices...);
+   }
+
+   template <typename I>
+   void setup_indices(uint64_t index_name, bool is_named, I index) {
+      if (is_named) {
+         eosio::check(index->index_name.value > 0, "All indices must be named if one is named.");
+      } else {
+         eosio::check(index->index_name.value <= 0, "All indices must be named if one is named.");
+         index->index_name = eosio::name{index_name};
+      }
+
+      index->contract_name = contract_name;
+      index->table_name = table_name;
+      index->tbl = this;
+
+      index->setup();
+      secondary_indices.push_back(index);
+   }
+
+   template <typename PrimaryIndex, typename... SecondaryIndices>
+   void init(eosio::name contract, eosio::name table, eosio::name db, PrimaryIndex prim_index, SecondaryIndices... indices) {
+      contract_name = contract;
+      table_name = table;
+      db_name = db.value;
+
+      bool is_named = false;
+      uint64_t index_name = 1;
+
+      primary_index = prim_index;
+      primary_index->contract_name = contract_name;
+      primary_index->table_name = table_name;
+      primary_index->tbl = this;
+
+      if (primary_index->index_name.value > 0) {
+         is_named = true;
+      } else {
+         primary_index->index_name = eosio::name{index_name};
+         ++index_name;
+      }
+
+      primary_index->setup();
+
+      if constexpr (sizeof...(indices) > 0) {
```

Comment:
It would be cleaner to define an overload of setup_indices that takes no indices.  This would also simplify your recursion.